gRPC master. Python tests

1. dbg on linux
https://grpc-testing.appspot.com/job/gRPC_master/config=dbg,language=python,platform=linux/8345/consoleFull

2. dbg on macos

https://grpc-testing.appspot.com/job/gRPC_master/config=dbg,language=python,platform=macos/8345/consoleFul

```
[gw2] FAILED grpc_protoc_plugin/beta_python_plugin_test.py::PythonPluginTest::testStreamingOutputCallCancelled...
[gw5] FAILED grpc_test/_adapter/_event_invocation_synchronous_event_service_test.py::EventInvocationSynchronousEventServiceTest::testSuccessfulStreamRequestStreamResponse <- grpc_test/framework/face/testing/event_invocation_synchronous_event_service_test_case.py 
...
[gw5] FAILED grpc_test/_adapter/_event_invocation_synchronous_event_service_test.py::EventInvocationSynchronousEventServiceTest::testSuccessfulUnaryRequestStreamResponse <- grpc_test/framework/face/testing/event_invocation_synchronous_event_service_test_case.py 
...
[gw4] FAILED grpc_test/_adapter/_future_invocation_asynchronous_event_service_test.py::FutureInvocationAsynchronousEventServiceTest::testSuccessfulStreamRequestStreamResponse <- grpc_test/framework/face/testing/future_invocation_asynchronous_event_service_test_case.py 
...
[gw0] FAILED grpc_test/early_adopter/implementations_test.py::EarlyAdopterImplementationsTest::testStreamStream 
...

```
3. Opt test on macos
https://grpc-testing.appspot.com/job/gRPC_master/config=opt,language=python,platform=macos/8345/consoleFull

```
...
[gw7] FAILED grpc_test/_adapter/_future_invocation_asynchronous_event_service_test.py::FutureInvocationAsynchronousEventServiceTest::testSuccessfulStreamRequestStreamResponse <- grpc_test/framework/face/testing/future_invocation_asynchronous_event_service_test_case.py 
[gw6] FAILED grpc_test/_adapter/_event_invocation_synchronous_event_service_test.py::EventInvocationSynchronousEventServiceTest::testSuccessfulStreamRequestStreamResponse <- grpc_test/framework/face/testing/event_invocation_synchronous_event_service_test_case.py 
...
[gw1] FAILED grpc_test/early_adopter/implementations_test.py::EarlyAdopterImplementationsTest::testStreamStream 
...
=================================== FAILURES ===================================
 FutureInvocationAsynchronousEventServiceTest.testSuccessfulStreamRequestStreamResponse 
[gw7] darwin -- Python 2.7.6 /jenkins/workspace/gRPC_master/config/opt/language/python/platform/macos/python2.7_virtual_environment/bin/python2.7
self = <grpc_test._adapter._future_invocation_asynchronous_event_service_test.FutureI...nchronousEventServiceTest testMethod=testSuccessfulStreamRequestStreamResponse>

    def testSuccessfulStreamRequestStreamResponse(self):
      for name, test_messages_sequence in (
          self.digest.stream_stream_messages_sequences.iteritems()):
        for test_messages in test_messages_sequence:
          requests = test_messages.requests()
          request_iterator = _PauseableIterator(iter(requests))
    
          # Use of a paused iterator of requests allows us to test that control is
          # returned to calling code before the iterator yields any requests.
          with request_iterator.pause():
            response_iterator = self.stub.inline_stream_in_stream_out(
                name, request_iterator, test_constants.SHORT_TIMEOUT)
>         responses = list(response_iterator)

grpc_test/framework/face/testing/future_invocation_asynchronous_event_service_test_case.py:158: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../python2.7_virtual_environment/lib/python2.7/site-packages/grpc/framework/face/_calls.py:118: in next
    return next(self._rendezvous)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <grpc.framework.face._control.Rendezvous object at 0x110b8e710>

    def next(self):
      with self._condition:
        while ((self._abortion is None) and
               (not self._values) and
               (not self._values_completed)):
          self._condition.wait()
        if self._abortion is not None:
>         raise _abortion_outcome_to_exception(self._abortion)
E         ExpirationError
...


```

--- a/src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pxd
+++ b/src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pxd
@@ -37,3 +37,5 @@ cdef class CompletionQueue:
   cdef bint is_polling
   cdef bint is_shutting_down
   cdef bint is_shutdown
+
+  cdef _interpret_event(self, grpc.grpc_event event)

--- a/src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx
+++ b/src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx
@@ -46,35 +46,13 @@ cdef class CompletionQueue:
     self.poll_condition = threading.Condition()
     self.is_polling = False
 
-  def poll(self, records.Timespec deadline=None):
-    # We name this 'poll' to avoid problems with CPython's expectations for
-    # 'special' methods (like next and __next__).
-    cdef grpc.gpr_timespec c_deadline = grpc.gpr_inf_future(
-        grpc.GPR_CLOCK_REALTIME)
+  cdef _interpret_event(self, grpc.grpc_event event):
     cdef records.OperationTag tag = None
     cdef object user_tag = None
     cdef call.Call operation_call = None
     cdef records.CallDetails request_call_details = None
     cdef records.Metadata request_metadata = None
     cdef records.Operations batch_operations = None
-    if deadline is not None:
-      c_deadline = deadline.c_time
-    cdef grpc.grpc_event event
-
-    # Poll within a critical section
-    # TODO consider making queue polling contention a hard error to enable
-    # easier bug discovery
-    with self.poll_condition:
-      while self.is_polling:
-        self.poll_condition.wait(float(deadline) - time.time())
-      self.is_polling = True
-    with nogil:
-      event = grpc.grpc_completion_queue_next(
-          self.c_completion_queue, c_deadline, NULL)
-    with self.poll_condition:
-      self.is_polling = False
-      self.poll_condition.notify()
-
     if event.type == grpc.GRPC_QUEUE_TIMEOUT:
       return records.Event(
           event.type, False, None, None, None, None, False, None)
@@ -103,6 +81,54 @@ cdef class CompletionQueue:
           event.type, event.success, user_tag, operation_call,
           request_call_details, request_metadata, tag.is_new_request,
           batch_operations)
+
+  def poll(self, records.Timespec deadline=None):
+    # We name this 'poll' to avoid problems with CPython's expectations for
+    # 'special' methods (like next and __next__).
+    cdef grpc.gpr_timespec c_deadline = grpc.gpr_inf_future(
+        grpc.GPR_CLOCK_REALTIME)
+    if deadline is not None:
+      c_deadline = deadline.c_time
+    cdef grpc.grpc_event event
+
+    # Poll within a critical section
+    # TODO(atash) consider making queue polling contention a hard error to
+    # enable easier bug discovery
+    with self.poll_condition:
+      while self.is_polling:
+        self.poll_condition.wait(float(deadline) - time.time())
+      self.is_polling = True
+    with nogil:
+      event = grpc.grpc_completion_queue_next(
+          self.c_completion_queue, c_deadline, NULL)
+    with self.poll_condition:
+      self.is_polling = False
+      self.poll_condition.notify()
+    return self._interpret_event(event)
+
+  def pluck(self, records.OperationTag tag, records.Timespec deadline=None):
+    # Plucking a 'None' tag is equivalent to passing control to GRPC core until
+    # the deadline.
+    cdef grpc.gpr_timespec c_deadline = grpc.gpr_inf_future(
+        grpc.GPR_CLOCK_REALTIME)
+    if deadline is not None:
+      c_deadline = deadline.c_time
+    cdef grpc.grpc_event event
+
+    # Poll within a critical section
+    # TODO(atash) consider making queue polling contention a hard error to
+    # enable easier bug discovery
+    with self.poll_condition:
+      while self.is_polling:
+        self.poll_condition.wait(float(deadline) - time.time())
+      self.is_polling = True
+    with nogil:
+      event = grpc.grpc_completion_queue_pluck(
+          self.c_completion_queue, <cpython.PyObject *>tag, c_deadline, NULL)
+    with self.poll_condition:
+      self.is_polling = False
+      self.poll_condition.notify()
+    return self._interpret_event(event)
 
   def shutdown(self):
     grpc.grpc_completion_queue_shutdown(self.c_completion_queue)

--- a/src/python/grpcio/grpc/_cython/_cygrpc/grpc.pxd
+++ b/src/python/grpcio/grpc/_cython/_cygrpc/grpc.pxd
@@ -292,6 +292,9 @@ cdef extern from "grpc/grpc.h":
   grpc_event grpc_completion_queue_next(grpc_completion_queue *cq,
                                         gpr_timespec deadline,
                                         void *reserved) nogil
+  grpc_event grpc_completion_queue_pluck(grpc_completion_queue *cq, void *tag,
+                                         gpr_timespec deadline,
+                                         void *reserved) nogil
   void grpc_completion_queue_shutdown(grpc_completion_queue *cq)
   void grpc_completion_queue_destroy(grpc_completion_queue *cq)
 

--- a/src/python/grpcio/grpc/_cython/_cygrpc/server.pyx
+++ b/src/python/grpcio/grpc/_cython/_cygrpc/server.pyx
@@ -89,6 +89,8 @@ cdef class Server:
     self.register_completion_queue(self.backup_shutdown_queue)
     self.is_started = True
     grpc.grpc_server_start(self.c_server)
+    # Ensure the core has gotten a chance to do the start-up work
+    self.backup_shutdown_queue.pluck(None, records.Timespec(None))
 
   def add_http2_port(self, address,
                      credentials.ServerCredentials server_credentials=None):

